<launch>


  <!-- Name of the map to use and initial position -->
  <arg name="map_file"       default="$(find second_project)/maps/map.yaml"/>
  <arg name="world_file"     default="$(find second_project)/maps/stage/map.world"/>
  <arg name="initial_pose_x" default="1.0"/>
  <arg name="initial_pose_y" default="1.0"/>
  <arg name="initial_pose_a" default="0.0"/>

  <arg name="odom_frame_id"   default="odom"/>
  <arg name="global_frame_id" default="map"/>

  <param name="/use_sim_time" value="true"/>

<!--  ****** Maps *****  -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)">
    <param name="frame_id" value="map"/>
  </node>

 <!--  ***************** Robot simulation *****************  -->
  <node pkg="stage_ros" type="stageros" name="stageros" args="$(arg world_file)">
    <param name="base_watchdog_timeout" value="0.5"/>
    <remap from="base_scan" to="scan"/>
  </node>


  <!--  ***************** Robot Model ***************** -->
  <!-- Parametri per le dimensioni del marker -->
  <arg name="marker_size_x" default="0.612"/>
  <arg name="marker_size_y" default="0.580"/>
  <arg name="marker_size_z" default="0.245"/>

  <!-- Node for publishing the cube marker -->
  <node pkg="second_project" name="marker" type="marker" output="screen">
    <!-- Passa i parametri del marker come argomenti -->
    <param name="marker_size_x" value="$(arg marker_size_x)"/>
    <param name="marker_size_y" value="$(arg marker_size_y)"/>
    <param name="marker_size_z" value="$(arg marker_size_z)"/>
  </node>

  <!--  ************** Goal Pub  ***************  -->

   <node name="goal_publisher" pkg="second_project" type="goal_publisher" output="screen">
        <param name="csv_file" value="$(find second_project)/waypoints.csv" />
   </node>

  <!--  ************** Navigation  ***************  -->
   <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam file="$(find second_project)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find second_project)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find second_project)/config/local_costmap_params.yaml" command="load"/>
    <rosparam file="$(find second_project)/config/global_costmap_params.yaml" command="load"/>
    <rosparam file="$(find second_project)/config/teb_local_planner_params.yaml" command="load" />
    <rosparam file="$(find second_project)/config/move_base_params.yaml" command="load" />

    <param name="base_global_planner" value="navfn/NavfnROS" />  <!--   planner  navfn/NavfnROS alternative  carrot_planner/CarrotPlanner -->
    <param name="planner_frequency" value="1.0" />
    <param name="planner_patience" value="5.0" /> <!-- How long the planner will wait in seconds in an attempt to find a valid plan before space-clearing operations are performed.  -->

    <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
    <param name="controller_frequency" value="5.0" />
    <param name="controller_patience" value="2.0" /> <!-- How long the controller will wait in seconds without receiving a valid control before space-clearing operations are performed.  -->

    <param name="clearing_rotation_allowed" value="true"/>
  </node>



  <!--  ****** amcl *****  -->
  <include file="$(find second_project)/launch/amcl.launch.xml">
    <arg name="scan_topic" value="scan"/>
    <arg name="use_map_topic" value="true"/>
    <arg name="initial_pose_x" value="$(arg initial_pose_x)"/>
    <arg name="initial_pose_y" value="$(arg initial_pose_y)"/>
    <arg name="initial_pose_a" value="$(arg initial_pose_a)"/>
  </include>

  <!--  **************** Visualisation **************** -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find second_project)/rviz/second_rviz_config.rviz"/>


</launch>